### ISSUE TITLE
analytical query object->subject joins

### ISSUE THREAD

**Issue by dpetran** [fluree/db#228]
This does not work, and should.

```
{
    "select": {
        "?s": ["*", {"follows": ["*"]}]
    },
    "where": [
        ["?s", "follows", "?o"],
        ["?o", "fullName", "?v"]
    ]
}
```

```
{
    "select": {
        "?s": ["*", {"follows": ["*"]}]
    },
    "where": [
        ["?s", "follows", "?o"],
        ["?o", "favNums", 3]
    ]
}
```

**Comment by bplatz** [comment]
Andrew J. has provided an example query that is not working to test this.

**Comment by mpoffald** [comment]
Tl;dr The join-related bugs seem to be artifacts of the current query pipeline architecture, and I propose doing nothing about any of this becasue we're rewriting it anyway.


1. Missing object logic, lost context when selecting

Example query:
```clj
{:context {:ex "http://example.org/ns/"}
 :select {'?s ["*" {:ex/friend ["*"]}]}
 :where [['?o :schema/name "Alice"]
        ['?s :ex/friend '?o]]}
``` 
Before I tried to fix anything this returned `[]`, because of a (known) limitation in the `compound` ns: there was no case for the second clause, in which the `s` id is not known, but we do have an `o`.

Once I addressed that, this would return data, but in a way that ignored the first clause limiting the results to friends named Alice. That is, it returned a crawl of everything about `s`s who have friends, and a crawl of _all_ of those friends. 

This is because of how the current pipeline streams limited data from one clause to the next:
- the first clause returns the subject id for Alice
- the second clause returns the `s` of whoever is friends with Alice
- `s` gets passed to our `select` statement handling, which just does [a crawl of the graph with that id and the `:spot` index](https://github.com/fluree/db/blob/main/src/fluree/db/query/fql.cljc#L34), with no knowledge of the restriction imposed by the join.

2. Clause order sensitivity, `flake-x-form` creation
This is the above example but with the `where` clauses reversed:
```clj
{:context {:ex "http://example.org/ns/"}
 :select {'?s ["*" {:ex/friend ["*"]}]}
 :where [['?s :ex/friend '?o]
        ['?o :schema/name "Alice"]]}
``` 

Initially this just broke with an exception, because the calculation of the `flake-x-form` fn was `nil`, which blew up when it was applied as a function later.

This was because the second clause does not share any vars with the `select` statement, and the calculation of the `flake-x-form` fn assumes it will (current parsing is order-sensitive). There was no case for this situation, so it fell through to `nil`.

I tried some fixes in which I introduced a concept of `join-vars`, either as a seprate concept or amending them to the `out-vars`, in order to get a (seemingly) sensible non-`nil` fn for `flake-x-form`. These attempts got me past that bug but led to downstream weirdness in results, because the combination of `flake-x-form` + `passthrough-fn` + the `select` statement logic. 

I abandoned that effort once we decided we were redoing the query pipeline anyway. 

Notably, this fix would've been both kind of invasive and different than the fix for the other clause order, which is the sort of many-special-custom-fixes situation we're trying to get away from with our rewrite.


### Some potentially useful takeaways
- Right now our subject comparator assumes longs, and is not prepared to handle variables in the subject position. I got past this in my experiments by rewriting `cmp-subj` to be:
```clj
(if (or (map? s1) (map? s2))
             0
             (cmp-long s2 s1))
```
but I'm not sure that's actually correct in all cases, I was just trying to get past an error in a specific case. So I don't propose making this change now, more an FYI for the rewrite.
- I did briefly think the issue was deeper than the query processing pipeline (at the flake/index level), but upon further investigation I realized my mental model was just wrong. I'm now fairly certain it's all just due to how we parse/execute queries -- or at least, if there are deeper issues, they were not discernible to me with the current query pipeline. 
- There are no tests anywhere for these cases. We'll want some for the rewrite.


@zonotope These look to me like issues that'll get resolved with the rewrite, if you agree then I propose we close this. 

I did write a couple of tests in the course of this debugging adventure, though, and I've pushed them up to the branch `update/join-tests` ([this commit](https://github.com/fluree/db/commit/8535e7872fbb7fccc47e874f96c202c959541818)), in case those are useful to your work. I won't PR them though, since they just fail.

**Comment by bplatz** [comment]
The `cmp-subj` should be avoided, as this is foundational to indexing and lookups and adds more work that will impact the performance of the whole system... so really an alternative approach would be preferred.

This statement is really a filtering function that requires an async lookup - which is something that is going to need to be supported regardless of this scenario or to support other filter functions that might require async activity.

The the approach will need to be, regardless of where this happens, to recognize this as an -opposite- requirement for what is already supported in `optional`. In `optional`, where instead of filtering, we recognize the absence of a value and substitute in `nil`. In this case that same recognition should occur but instead of subbing in `nil`, it should drop the result.

**Comment by bplatz** [comment]
@mpoffald I just pushed a fix for this onto this branch: https://github.com/fluree/db/tree/fix/subject-filtering

You can take a look and either go with it (and write a test!), or just use it to see what the fix involved.

I didn't fully test this across different scenarios but it should work!

**Comment by mpoffald** [comment]
@bplatz what portion of the issue is this intended to fix? I just pulled it down and ran my join tests and they did not pass.

**Comment by mpoffald** [comment]
🤔 That's odd, my test are similar and don't pass. Could you push some passing tests?

**Comment by bplatz** [comment]
@mpoffald Here is the data I used which I had handy... only tested it with this:
```
  [{:id               :ex/alice,
    :type             :ex/User,
    :schema/name      "Alice"
    :ex/location      {:ex/state   "NC"
                       :ex/country "USA"}}
   {:id               :ex/john,
    :type             :ex/User,
    :schema/name      "John"
    :ex/location      {:ex/state   "MI"
                       :ex/country "USA"}}]
```
queries:
```
  @(fluree/query db {:select {'?s [:* ]}
                     :where  [['?s :ex/location '?o]
                              ['?o :ex/state "NC"]]})
```
==>> only returns alice

```
  @(fluree/query db {:select {'?s [:* ]}
                     :where  [['?s :ex/location '?o]
                              ['?o :ex/state "MI"]]})
```
==> only returns John
```
  @(fluree/query db {:select {'?s [:* ]}
                     :where  [['?s :ex/location '?o]
                              ['?o :ex/state "SC"]]})
```
==> Returns nobody

**Comment by mpoffald** [comment]
Yes, those examples do work, thanks!  But they don't use quite the same select mechanism as the original query that was reported , and I believe the issue I described in part 1 still applies. 

On your branch, if I do:
```clj
(let [conn   (test-utils/create-conn)
      ledger @(fluree/create conn "query/compound")
      db     @(fluree/stage
               ledger
               [{ :context {:ex "http://example.org/ns/"}
                 :id               :ex/alice,
                 :type             :ex/User,
                 :schema/name      "Alice"
                 :ex/location      {:ex/state   "NC"
                                    :ex/country "USA"}
                 :ex/friend [:ex/john :ex/brian]}
                {:context {:ex "http://example.org/ns/"}
                 :id               :ex/john,
                 :type             :ex/User,
                 :schema/name      "John"
                 :ex/location      {:ex/state   "MI"
                                    :ex/country "USA"}}
                {:context {:ex "http://example.org/ns/"}
                 :id               :ex/brian
                 :type             :ex/User,
                 :schema/name      "Brian"
                 :ex/location      {:ex/state   "MI"
                                    :ex/country "USA"}}])]
  @(fluree/query db {:context {:ex "http://example.org/ns/"}
                     :select {'?s [:*, {:ex/friend [:*]}]}
                     :where  [['?s :ex/friend '?o]
                              ['?o :schema/name "John"]]}) )
```

I get back
```clj
[{:id :ex/alice,
  :rdf/type [:ex/User],
  :schema/name "Alice",
  :ex/location {:id "_:f211106232532993"},
  :ex/friend
  [{:id :ex/john,
    :rdf/type [:ex/User],
    :schema/name "John",
    :ex/location {:id "_:f211106232532996"}}
   {:id :ex/brian,
    :rdf/type [:ex/User],
    :schema/name "Brian",
    :ex/location {:id "_:f211106232532997"}}]}]
```
But Brian should not be in that result. 

Taking a step back here though, my understanding was that we had agreed that if this bug was in the query pipeline (and not somewhere deeper, like indexing) then we were going to punt on this until after the rewrite. If @zonotope agrees that our upcoming refactoring will address these issues, then I stand by my proposal to close this issue.

However, if you'd really like to propose a fix with the existing pipeline, I'd like to see tests like this example (or the ones I pushed [here](https://github.com/fluree/db/compare/main...update/join-tests)) working on the branch, and we can talk about this in a PR.

**Comment by bplatz** [comment]
@mpoffald As I mentioned in my initial comment, I pushed this up knowing it might not get used, it was only a few lines of code so if tossed no big deal. I thought it would be useful to know what the fix looked like.

For this reason I wasn't exhaustive with it. If we do end up using it, happy to pair up to cover the tests you mention - which I haven't looked at yet.

**Comment by mpoffald** [comment]
Understood, and thanks for the offer! If we decide to proceed with the current query pipeline rather than go with the refactor, we can pick this back up and finish it.

**Comment by bplatz** [comment]
@mpoffald I just looked at your provided query, and it is returning the correct results.
```
  @(fluree/query db {:context {:ex "http://example.org/ns/"}
                     :select {'?s [:*, {:ex/friend [:*]}]}
                     :where  [['?s :ex/friend '?o]
                              ['?o :schema/name "John"]]}) )
```
The query will find only `:ex/alice`, which it does correctly. You then do a graph crawl in your `:select` which asks it to retrieve all friends of `:ex/alice`, which includes both John and Brian.

**Comment by mpoffald** [comment]
Interesting, that's good to know! I definitely would've expected the first clause to restrict the results from which the `select` statement selects. This behavior feels unintuitive to me, but that's outside the scope of this discussion. 😅 

It sounds like that works as expected, then. However, at least when I run this locally, it returns no results when I reorder the `where` clauses, so that's still an issue. But again, something for the future, if we decide to keep building on the current query pipeline.
